docs: 🤗 Hugging Face Transformers API compatibility (#1146)
NVIDIA-NeMo/Automodel: docs/guides/huggingface-api-compatibility.md

# 🤗 Hugging Face Transformers API compatibility

NeMo Automodel is built to work with the 🤗 Hugging Face ecosystem.
In practice, compatibility comes in two layers:

- **API compatibility**: for many workflows you can keep your existing `transformers` code and swap in NeMo Automodel “drop-in” wrappers (`NeMoAutoModel*`, `NeMoAutoTokenizer`) with minimal changes.
- **Artifact compatibility**: NeMo Automodel produces **Hugging Face–compatible checkpoints** (config + tokenizer + safetensors) that can be loaded by Hugging Face Transformers and downstream tools (vLLM, SGLang, etc.).

> This page summarizes what “HF compatibility” means in NeMo Automodel, calls out differences you should be aware of, and provides side-by-side examples.

Reviewer: ```suggestion
This page summarizes what "HF compatibility" means in NeMo Automodel, calls out differences you should be aware of, and provides side-by-side examples.
```

> ## Transformers version compatibility: v4 and v5

Reviewer: ```suggestion
## Transformers Version Compatibility: v4 and v5
```

### Transformers v4 (current default)

NeMo Automodel currently pins Hugging Face Transformers to the **v4** major line (see `pyproject.toml`, currently `transformers<=4.57.5`).

This means:

- NeMo Automodel is primarily tested and released against **Transformers v4.x**
- New model releases on the Hugging Face Hub that require a newer Transformers may require upgrading NeMo Automodel as well (similar to upgrading `transformers` directly)

> ### Transformers v5 (forward-compatibility + checkpoint interoperability)

Reviewer: ```suggestion
### Transformers v5 (Forward-Compatibility + Checkpoint Interoperability)
```

NeMo Automodel addresses this in two complementary ways:

- **Forward-compatibility shims**: NeMo Automodel includes small compatibility patches to smooth over known API differences across Transformers releases (for example, cache utility method names). The built-in recipes apply these patches automatically.
- **Backports where needed**: for some model families, NeMo Automodel may vendor/backport Hugging Face code that originated in the v5 development line so users can run those models while staying on a pinned v4 dependency.
- **Stable artifact format**: NeMo Automodel checkpoints are written in Hugging Face–compatible `save_pretrained` layouts (config + tokenizer + safetensors). These artifacts are designed to be loadable by both Transformers **v4** and **v5** (and non-Transformers tools that consume HF-style model repos).

:::{note}
If you are running Transformers v5 in another environment, you can still use NeMo Automodel–produced consolidated checkpoints with Transformers’ standard loading APIs. For details on the checkpoint layouts, see [checkpointing](checkpointing.md).
:::

> ## What’s “drop-in” vs what’s different

Reviewer: ```suggestion
## Drop-In Compatibility and Key Differences
```

> ### Drop-in (same mental model as Transformers)

Reviewer: ```suggestion
### Drop-In (Same Mental Model as Transformers)
```

- **Load by model ID or local path**: `from_pretrained(...)`
- **Standard HF config objects**: `AutoConfig` / `config.json`
- **Tokenizers**: standard `PreTrainedTokenizerBase` behavior, including `__call__` to create tensors and `decode`/`batch_decode`
- **Generation**: `model.generate(...)` and the usual generation kwargs

> ### Differences (where NeMo Automodel adds value or has constraints)

Reviewer: ```suggestion
### Differences (Where NeMo Automodel Adds Value or Has Constraints)
```

- **Performance features**: NeMo Automodel can automatically apply optional kernel patches/optimizations (e.g., SDPA selection, Liger kernels, DeepEP, etc) while keeping the public model API the same.

> - **Distributed training stack**: NeMo Automodel’s recipes/CLI are designed for multi-GPU/multi-node fine-tuning with PyTorch-native distributed features (FSDP2, pipeline parallelism, etc.).

Reviewer: ```suggestion
- **Distributed training stack**: NeMo Automodel's recipes/CLI are designed for multi-GPU/multi-node fine-tuning with PyTorch-native distributed features (FSDP2, pipeline parallelism, etc.).
```

      <td><code>model.save_pretrained(path, checkpointer=...)</code></td>
    </tr>
  </tbody>
</table>
:::

## Side-by-side examples

### Load a model and tokenizer (Transformers v4)

:::{raw} html
<table>
  <thead>
    <tr>
      <th style="width: 50%;">🤗 Hugging Face (<code>transformers</code>)</th>
      <th style="width: 50%;">NeMo Automodel (<code>nemo_automodel</code>)</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td style="vertical-align: top;">
        <div class="highlight"><pre><code>import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model_id = "gpt2"

tokenizer = AutoTokenizer.from_pretrained(model_id)
model = AutoModelForCausalLM.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
)</code></pre></div>
      </td>
      <td style="vertical-align: top;">
        <div class="highlight"><pre><code>import torch
from nemo_automodel import NeMoAutoModelForCausalLM, NeMoAutoTokenizer

model_id = "gpt2"

tokenizer = NeMoAutoTokenizer.from_pretrained(model_id)
model = NeMoAutoModelForCausalLM.from_pretrained(
    model_id,
    torch_dtype=torch.bfloat16,
)</code></pre></div>
      </td>
    </tr>
  </tbody>
</table>
:::

> ### Text generation

Reviewer: ```suggestion
### Text Generation
```

This snippet assumes you already have a `model` and `tokenizer` (see the loading snippet above).

:::{raw} html
<table>
  <thead>
    <tr>

>       <th style="width: 50%;">🤗 Hugging Face (<code>transformers</code>)</th>

Reviewer: ```suggestion
      <th style="width: 50%;">Hugging Face (<code>transformers</code>)</th>
```

    <tr>
      <td style="vertical-align: top; padding-top: 0;">
        <div class="highlight" style="margin-top: 0;"><pre style="margin: 0;"><code>import torch

prompt = "Write a haiku about GPU kernels."
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

with torch.inference_mode():
    out = model.generate(**inputs, max_new_tokens=64)

print(tokenizer.decode(out[0], skip_special_tokens=True))</code></pre></div>
      </td>
      <td style="vertical-align: top; padding-top: 0;">
        <div class="highlight" style="margin-top: 0;"><pre style="margin: 0;"><code>import torch

prompt = "Write a haiku about GPU kernels."
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

with torch.inference_mode():
    out = model.generate(**inputs, max_new_tokens=64)

print(tokenizer.decode(out[0], skip_special_tokens=True))</code></pre></div>
      </td>
    </tr>
  </tbody>
</table>
:::

### Tokenizers (Transformers vs NeMo Automodel)

NeMo Automodel provides `NeMoAutoTokenizer` as a Transformers-like auto-tokenizer with a small registry for specialized backends (and a safe fallback when no specialization is needed).

:::{raw} html
<table>
  <thead>
    <tr>

>       <th style="width: 50%;">🤗 Hugging Face (<code>transformers</code>)</th>

Reviewer: ```suggestion
      <th style="width: 50%;">Hugging Face (<code>transformers</code>)</th>
```

# 🤗 Hugging Face Transformers API Compatibility

NeMo Automodel is built to work with the 🤗 Hugging Face ecosystem.
In practice, compatibility comes in two layers:

> - **API compatibility**: for many workflows, you can keep your existing `transformers` code and swap in NeMo Automodel “drop-in” wrappers (`NeMoAutoModel*`, `NeMoAutoTokenizer`) with minimal changes.

Reviewer: ```suggestion
- **API compatibility**: for many workflows, you can keep your existing `transformers` code and swap in NeMo AutoModel “drop-in” wrappers (`NeMoAutoModel*`, `NeMoAutoTokenizer`) with minimal changes.
```

- **Artifact compatibility**: NeMo Automodel produces **Hugging Face-compatible checkpoints** (config + tokenizer + safetensors) that can be loaded by Hugging Face Transformers and downstream tools (vLLM, SGLang, etc.).

This page summarizes what "HF compatibility" means in NeMo Automodel, calls out differences you should be aware of, and provides side-by-side examples.

## Transformers Version Compatibility: v4 and v5

### Transformers v4 (Current Default)

> NeMo Automodel currently pins Hugging Face Transformers to the **v4** major line (see `pyproject.toml`, currently `transformers<=4.57.5`).

Reviewer: ```suggestion
NeMo AutoModel currently pins Hugging Face Transformers to the **v4** major line (see `pyproject.toml`, currently `transformers<=4.57.5`).
```

This means:

> - NeMo Automodel is primarily tested and released against **Transformers v4.x**

Reviewer: ```suggestion
- NeMo AutoModel is primarily tested and released against **Transformers v4.x**
```

- New model releases on the Hugging Face Hub that require a newer Transformers may require upgrading NeMo Automodel as well (similar to upgrading `transformers` directly)

### Transformers v5 (Forward-Compatibility and Checkpoint Interoperability)

Transformers **v5** introduces breaking changes across some internal utilities (e.g., cache APIs) and adds/reshapes tokenizer backends for some model families.

NeMo Automodel addresses this in two complementary ways:

> - **Forward-compatibility shims**: NeMo Automodel includes small compatibility patches to smooth over known API differences across Transformers releases (for example, cache utility method names). The built-in recipes apply these patches automatically.

Reviewer: ```suggestion
- **Forward-compatibility shims**: NeMo AutoModel includes small compatibility patches to smooth over known API differences across Transformers releases (for example, cache utility method names). The built-in recipes apply these patches automatically.
```

- **Backports where needed**: for some model families, NeMo Automodel may vendor/backport Hugging Face code that originated in the v5 development line so users can run those models while staying on a pinned v4 dependency.

> - **Stable artifact format**: NeMo Automodel checkpoints are written in Hugging Face–compatible `save_pretrained` layouts (config + tokenizer + safetensors). These artifacts are designed to be loadable by both Transformers **v4** and **v5** (and non-Transformers tools that consume HF-style model repos).

Reviewer: ```suggestion
- **Stable artifact format**: NeMo AutoModel checkpoints are written in Hugging Face-compatible `save_pretrained` layouts (config + tokenizer + safetensors). These artifacts are designed to be loadable by both Transformers **v4** and **v5** (and non-Transformers tools that consume HF-style model repos).
```

:::{note}
If you are running Transformers v5 in another environment, you can still use NeMo Automodel-produced consolidated checkpoints with Transformers' standard loading APIs. For details on the checkpoint layouts, see [checkpointing](checkpointing.md).
:::

## Drop-In Compatibility and Key Differences

### Drop-In (Same Mental Model as Transformers)

- **Load by model ID or local path**: `from_pretrained(...)`
- **Standard HF config objects**: `AutoConfig` / `config.json`
- **Tokenizers**: standard `PreTrainedTokenizerBase` behavior, including `__call__` to create tensors and `decode`/`batch_decode`
- **Generation**: `model.generate(...)` and the usual generation kwargs

### Differences (Where NeMo Automodel Adds Value or Has Constraints)

- **Performance features**: NeMo Automodel can automatically apply optional kernel patches/optimizations (e.g., SDPA selection, Liger kernels, DeepEP, etc.) while keeping the public model API the same.

> - **Distributed training stack**: NeMo Automodel's recipes/CLI are designed for multi-GPU/multi-node fine-tuning with PyTorch-native distributed features (FSDP2, pipeline parallelism, etc.).

Reviewer: ```suggestion
- **Distributed training stack**: NeMo AutoModel's recipes/CLI are designed for multi-GPU/multi-node fine-tuning with PyTorch-native distributed features (FSDP2, pipeline parallelism, etc.).
```

> - **CUDA expectation**: NeMo Automodel's `NeMoAutoModel*` wrappers are primarily optimized for NVIDIA GPU workflows, and offer support for CPU workflows as well.

Reviewer: ```suggestion
- **CUDA expectation**: NeMo AutoModel's `NeMoAutoModel*` wrappers are primarily optimized for NVIDIA GPU workflows, and offer support for CPU workflows as well.
```

:::{important}
`NeMoAutoModelForCausalLM.from_pretrained(...)` currently assumes CUDA is available (it uses `torch.cuda.current_device()` internally). If you need CPU-only inference, use Hugging Face `transformers` directly.
:::

> ## API Mapping (Transformers and NeMo Automodel)

Reviewer: ```suggestion
## API Mapping (Transformers and NeMo AutoModel)
```

prompt = "Write a haiku about GPU kernels."
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

with torch.inference_mode():
    out = model.generate(**inputs, max_new_tokens=64)

print(tokenizer.decode(out[0], skip_special_tokens=True))</code></pre></div>
      </td>
      <td style="vertical-align: top; padding-top: 0;">
        <div class="highlight" style="margin-top: 0;"><pre style="margin: 0;"><code>import torch

prompt = "Write a haiku about GPU kernels."
inputs = tokenizer(prompt, return_tensors="pt").to(model.device)

with torch.inference_mode():
    out = model.generate(**inputs, max_new_tokens=64)

print(tokenizer.decode(out[0], skip_special_tokens=True))</code></pre></div>
      </td>
    </tr>
  </tbody>
</table>
:::

### Tokenizers (Transformers vs NeMo Automodel)

NeMo Automodel provides `NeMoAutoTokenizer` as a Transformers-like auto-tokenizer with a small registry for specialized backends (and a safe fallback when no specialization is needed).

:::{raw} html
<table>
  <thead>
    <tr>
      <th style="width: 50%;">🤗 Hugging Face (<code>transformers</code>)</th>

>       <th style="width: 50%;">NeMo Automodel (<code>nemo_automodel</code>)</th>

Reviewer: ```suggestion
      <th style="width: 50%;">NeMo AutoModel (<code>nemo_automodel</code>)</th>
```
